luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -16,225 +16,521 @@ if not gadgetHandler:IsSyncedCode() then
 
-	local explosion = {
-		weaponDef          = 0, -- todo: Great. Now I gotta make a weapon for all of these.
-		edgeEffectiveness  = 1,
-		explosionSpeed     = 10000,
-		damageGround       = false,
-	}
+	local index = 1
+	while index <= sizeg do
+		local timedArea = group[index]
+		if time <= timedArea[1] then
+			local explosion = explosionCaches[timedArea[2]]
+			explosion.owner = timedArea[3] ~= -1 and timedArea[3] or nil 
+			spSpawnExplosion(timedArea[4], timedArea[5], timedArea[6], 0, 0, 0, explosion)
+		elseif index == sizeg then
+			group[index] = nil
+		else
+			group[index] = group[sizeg]
+			group[sizeg] = nil
+			index = index - 1
+			sizeg = sizeg - 1
+		end
+		index = index + 1
+	end
+end
 
-	-- Deal recurring damage by spawning explosions.
-	for index = #frameAreas, 1, -1 do
-		local timedArea = frameAreas[index]
-		if timedArea ~= empty then
-			if timedArea.endTime > frameTime then
-				-- todo: Will napalm/acid destroy shields now? Troubling thought.
-				explosion.weaponDef = timedArea.params.atd_wdef
-				Spring.SpawnExplosion(timedArea.x, timedArea.y, timedArea.z, 0, 0, 0, explosion)
-			else
-				StopTimedArea(frameAreas, index)
+local function cancelDelayedAreas(maxGameFrame)
+	local enabled, capacity, sx, sy, sz, dx, dy, dz, radiusTest
+	for shieldUnitID, radiusSq in pairs(shieldUnits) do
+		enabled, capacity = spGetUnitShieldState(shieldUnitID)
+		enabled = enabled and not spGetUnitIsStunned(shieldUnitID)
+		if enabled and capacity > 8 then
+			sx, sy, sz = spGetUnitPosition(shieldUnitID)
+			-- We don't care about the correctness of a short-lived, single-purpose array.
+			-- So when suppressing, we just replace a suppressed area's xpos with `false`.
+			for frame, delayedAreas in pairs(delayQueue) do
+				if frame <= maxGameFrame then
+					for ii = 1, #delayedAreas, 5 do
+						if delayedAreas[ii] then
+							dx = delayedAreas[ii  ] - sx
+							dy = delayedAreas[ii+1] - sy
+							dz = delayedAreas[ii+2] - sz
+							radiusTest = delayedAreas[ii+3].area_radius + radiusSq
+							if dx*dx + dz*dz < radiusTest and dy*dy < radiusTest then
+								delayedAreas[ii] = false
+							end
+						end
+					end
+				end
 			end
 		end
 	end
 end
 
 ---------------------------------------------------------------------------------------------------------------
+---- Gadget call-ins
 
 function gadget:Initialize()
-	for wdid, _ in pairs(WeaponAreaDefs) do
-		Script.SetWatchExplosion(wdid, true)
+	if not next(weaponTriggerParams) and not next(destroyTriggerParams) then
+		Spring.Log(gadget:GetInfo().name, LOG.INFO, "No timed area defs found. Removing gadget.")
+		gadgetHandler:RemoveGadget(self)
+	end
+
+	for weaponDefID, _ in pairs(weaponTriggerParams) do
+		Script.SetWatchExplosion(weaponDefID, true)
+	end
+
+	time  = 1 + (math.floor(Spring.GetGameFrame() / frameResolution))
+	frame = 1 + (math.floor(Spring.GetGameFrame() / frameResolution) % groupCount)
+	ticks = 1 + (Spring.GetGameFrame() % frameResolution)
```

Comment:
Aren't these all just 1, since `Initialize` is called on frame 0?

Reply:
/luarules reload